### PR TITLE
Add Dataset support to objectstorage

### PR DESCRIPTION
**Pull request by bolkedebruin**
This means you can now do ObjectStoragePath(a_dataset)

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by hussein-awala — APPROVED** [review]
Nice! LGTM

**Review by eladkal — APPROVED** [review]

**Review by dirrao — APPROVED** [review]

File: airflow/io/path.py [hunk 1]
```diff
@@ -129,7 +130,11 @@ def __new__(
 
             return _cls(_cls._format_parsed_parts(drv, root, parts, **other_kwargs), **new_kwargs)
 
-        url = stringify_path(other)
+        if isinstance(other, Dataset):
+            url = other.uri
+        else:
+            url = stringify_path(other)
```
Comment (uranusjr) [inline airflow/io/path.py]: I wonder if we should do this from the other direction, making Dataset a PathLike instead by implementing `__fspath__` and compatible with `stringify_path`.

Comment (bolkedebruin) [reply airflow/io/path.py]: Mmm Iike that too. And maybe both ways then?

**Comment by jedcunningham** [comment]
(@bolkedebruin, also moving this to 2.9)

**Pull request merged**